File: python/mxnet/executor_manager.py
# coding: utf-8
# pylint: disable=invalid-name, protected-access, too-many-locals, too-many-arguments, too-many-statements
"""Executor manager"""
from __future__ import absolute_import

import logging
import numpy as np

from .base import mx_real_t
from . import ndarray as nd
from .context import cpu


def _split_input_slice(batch_size, work_load_list):
    """Get input slice from the input shape.
    Parameters
    ----------
    batch_size : int
        The number of samples in a mini-batch.
    work_load_list : list of float or int, optional
        The list of work load for different devices,
        in the same order as ctx
    Returns
    -------
    slices : list of slice
        The split slices to get a specific slice.
    Raises
    ------
    ValueError
        If there are two many splits such that some slice can be empty.
    """
    total_work_load = sum(work_load_list)
    batch_num_list = [round(work_load * batch_size / total_work_load)
                      for work_load in work_load_list]
    batch_num_sum = sum(batch_num_list)
    if batch_num_sum < batch_size:
        batch_num_list[-1] += batch_size - batch_num_sum
    slices = []
    end = 0
    for batch_num in batch_num_list:
        begin = int(min((end, batch_size)))
        end = int(min((begin + batch_num, batch_size)))
        if begin >= end:
            raise ValueError('Too many slices such that some splits are empty')
        slices.append(slice(begin, end))
    return slices

def _check_arguments(symbol):
    """Check the argument names of symbol.
    This function checks the duplication of arguments in Symbol.
    The check is done for feedforward net for now.
    Parameters
    ----------
    symbol : Symbol
        The network configuration
    """
    arg_set = set()
    arg_names = symbol.list_arguments()
    for name in arg_names:
        if name in arg_set:
            raise ValueError(('Find duplicated argument name \"%s\", ' +
                              'please make the weight name non-duplicated(using name arguments), ' +
                              'arguments are %s') % (name, str(arg_names)))
        arg_set.add(name)

    aux_set = set()
    aux_names = symbol.list_auxiliary_states()
    for name in aux_names:
        if name in aux_set:
            raise ValueError(
                ('Find duplicated auxiliary param name \"%s\", ' +
                 'please make the weight name non-duplicated(using name arguments), ' +
                 'arguments are %s, auxiliary params are %s'
                ) % (name, str(arg_names), str(aux_names)))
        aux_set.add(name)

def _load_general(data, targets):
    """Load a list of arrays into a list of arrays specified by slices"""
    for d_src, d_targets in zip(data, targets):
        if isinstance(d_targets, nd.NDArray):
            d_src.copyto(d_targets)
        else:
            for slice_idx, d_dst in d_targets:
                d_src[slice_idx].copyto(d_dst)

def _load_data(batch, targets):
    """Load data into sliced arrays"""
    _load_general(batch.data, targets)

def _load_label(batch, targets):
    """Load label into sliced arrays"""
    _load_general(batch.label, targets)

# pylint: disable=too-many-branches
def _bind_exec(sym, ctx, input_shapes, param_names, need_grad=False,
               base_exec=None, shared_data_arrays=None, input_types=None, logger=logging):
    """bind executor for bucketing, potentially sharing data with an existing executor."""
    arg_shape, _, aux_shape = sym.infer_shape(**input_shapes)
    assert(arg_shape is not None)
    if input_types is None:
        input_types = {k: mx_real_t for k in input_shapes.keys()}
    arg_types, _, aux_types = sym.infer_type(**input_types)
    assert(arg_types is not None)

    arg_arrays = []
    grad_arrays = {} if need_grad != False else None

    arg_names = sym.list_arguments()

    if need_grad is False:
        need_grad = set()
    elif need_grad is True:
        need_grad = set(arg_names) - set(input_shapes.keys())
    elif isinstance(need_grad, set):
        pass
    else:
        raise AssertionError("need_grad must be boolean or set.")
    grad_req = {name:('write' if name in need_grad else 'null') for name in arg_names}


    # create or borrow arguments and gradients
    for i, name in enumerate(arg_names):
        if not name in param_names:
            # data or label
            if shared_data_arrays is not None and \
                    name in shared_data_arrays:
                arg_arr = shared_data_arrays[name]

                if np.prod(arg_arr.shape) >= np.prod(arg_shape[i]):
                    # good, we can share this memory
                    assert(arg_types[i] == arg_arr.dtype)
                    arg_arr = arg_arr.reshape(arg_shape[i])
                else:
                    logger.warning(('bucketing: data "%s" has a shape %s' % (name, arg_shape[i])) +
                                   (', which is larger than already allocated ') +
                                   ('shape %s' % (arg_arr.shape,)) +
                                   ('. Need to re-allocate. Consider putting ') +
                                   ('default_bucket_key to be the bucket taking the largest ') +
                                   ('input for better memory sharing.'))
                    arg_arr = nd.zeros(arg_shape[i], ctx, dtype=arg_types[i])

                    # replace existing shared array because the new one is bigger
                    shared_data_arrays[name] = arg_arr
            else:
                arg_arr = nd.zeros(arg_shape[i], ctx, dtype=arg_types[i])
                if shared_data_arrays is not None:
                    shared_data_arrays[name] = arg_arr

            arg_arrays.append(arg_arr)
        else:
            # model parameter
            if base_exec is None:
                arg_arr = nd.zeros(arg_shape[i], ctx, dtype=arg_types[i])
                if name in need_grad:
                    grad_arr = nd.zeros(arg_shape[i], ctx, dtype=arg_types[i])
                    grad_arrays[name] = grad_arr
            else:
                arg_arr = base_exec.arg_dict[name]
                assert arg_arr.shape == arg_shape[i]
                assert arg_arr.dtype == arg_types[i]
                if name in need_grad:
                    grad_arrays[name] = base_exec.grad_dict[name]
            arg_arrays.append(arg_arr)

    # create or borrow aux variables
    if base_exec is None:
        aux_arrays = [nd.zeros(s, ctx, dtype=t) for s, t in zip(aux_shape, aux_types)]
    else:
        for i, a in enumerate(base_exec.aux_arrays):
            assert aux_shape[i] == a.shape
            assert aux_types[i] == a.dtype

        aux_arrays = [a for a in base_exec.aux_arrays]

    executor = sym.bind(ctx=ctx, args=arg_arrays, args_grad=grad_arrays,
                        aux_states=aux_arrays,
                        grad_req=grad_req, shared_exec=base_exec)
    return executor

class DataParallelExecutorGroup(object):
    """A group of executors living on different devices, for data parallelization.

    Parameters
    ----------
    sym: Symbol
        The network configuration.
    arg_names: list of str
        Equals `sym.list_arguments()`
    param_names: list of str
        List of names of all trainable parameters.
    ctx: list of Context
        List of devices for training (data parallelization)
    slices: list of int
        Describes how the data parallelization splits data into different devices.
    train_data: DataIter (or DataBatch)
        The dataset for training. It could be any object with `provide_data` and
        `provide_label` properties. Loading of actual data is not necessarily needed
        at this stage.
    shared_grop: DataParallelExecutorGroup
        An existing executor group, if to share parameters with it.
    """
    def __init__(self, sym, arg_names, param_names, ctx, slices, train_data, shared_group=None):
        # make sure the architecture is valid
        _check_arguments(sym)

        if shared_group is None:
            self.shared_data_arrays = [{} for _ in ctx]
        else:
            self.shared_data_arrays = shared_group.shared_data_arrays

        self.data_names = [x[0] for x in train_data.provide_data]
        self.label_names = [x[0] for x in train_data.provide_label]
        self.aux_names = sym.list_auxiliary_states()
        self.param_idx = [i for i in range(len(arg_names)) if arg_names[i] in param_names]
        self.param_names = [arg_names[i] for i in self.param_idx]

        self.train_execs = []
        for i, ctxi in enumerate(ctx):
            data_shapes = {k: tuple([slices[i].stop-slices[i].start] + list(v[1:]))
                           for k, v in train_data.provide_data + train_data.provide_label}
            shared_exec = None if shared_group is None else shared_group.train_execs[i]
            train_exec = _bind_exec(sym, ctxi, data_shapes, self.param_names,
                                    need_grad=True, base_exec=shared_exec,
                                    shared_data_arrays=self.shared_data_arrays[i])
            self.train_execs.append(train_exec)

        # data structure
        self.data_arrays = [[(slices[i], e.arg_dict[name]) for i, e in enumerate(self.train_execs)]
                            for name in self.data_names]
        self.label_arrays = [[(slices[i], e.arg_dict[name]) for i, e in enumerate(self.train_execs)]
                             for name in self.label_names]

        self.param_arrays = [[e.arg_arrays[i] for e in self.train_execs]
                             for i in self.param_idx]
        self.grad_arrays = [[e.grad_arrays[i] for e in self.train_execs]
                            for i in self.param_idx]

        self.aux_arrays = [[e.aux_arrays[i] for e in self.train_execs]
                           for i in range(len(self.aux_names))]

        self.slices = slices

    def load_data_batch(self, data_batch):
        """ load data and labels into arrays """
        _load_data(data_batch, self.data_arrays)
        _load_label(data_batch, self.label_arrays)

    def forward(self, is_train=False):
        """ Perform a forward pass on each executor """
        for texec in self.train_execs:
            texec.forward(is_train=is_train)

    def backward(self):
        """ Perform a backward pass on each executor """
        for texec in self.train_execs:
            texec.backward()

    def update_metric(self, metric, labels):
        """ Update evaluation metric with label and current outputs """
        for texec, islice in zip(self.train_execs, self.slices):
            labels_slice = [label[islice] for label in labels]
            metric.update(labels_slice, texec.outputs)

class DataParallelExecutorManager(object):
    """ Helper class to manage multiple executors for data parallelism.
    Parameters
    ----------
    symbol : Symbol
        output symbol
    ctx : list of Context
        devices to run on
    param_names: list of str
        Name of all trainable parameters of the network.
    arg_names: list of str
        Name of all arguments of the network.
    aux_names: list of str
        Name of all auxiliary states of the network.
    train_data : DataIter
        Training data iterator.
    work_load_list : list of float or int, optional
        The list of work load for different devices,
        in the same order as ctx
    logger : logging logger
        When not specified, default logger will be used.
    sym_gen : a function that generate new Symbols depending on different
        input shapes. Used only for bucketing.
    """
    def __init__(self, symbol, ctx, train_data,
                 arg_names, param_names, aux_names,
                 work_load_list=None, logger=None, sym_gen=None):
        if logger is None:
            logger = logging
        # preparation
        num_device = len(ctx)
        logger.info('Start training with %s', str(ctx))

        if work_load_list is None:
            work_load_list = [1] * num_device
        assert isinstance(work_load_list, list) and len(work_load_list) == num_device, \
            "Invalid settings for work load. "

        slices = _split_input_slice(train_data.batch_size, work_load_list)
        self.slices = slices

        self.arg_names = arg_names
        self.param_names = param_names
        self.aux_names = aux_names
        self.ctx = ctx

        self.execgrp = DataParallelExecutorGroup(symbol, self.arg_names, self.param_names, self.ctx,
                                                 self.slices, train_data)
        self.symbol = symbol

        self.sym_gen = sym_gen
        self.curr_execgrp = None # this is set when data is loaded
        if self.sym_gen is not None:
            self.execgrp_bucket = {train_data.default_bucket_key: self.execgrp}


    def install_monitor(self, monitor):
        """ Install monitor on all executors """
        if self.sym_gen is not None:
            raise NotImplementedError("Monitoring is not implemented for bucketing")

        for train_exec in self.execgrp.train_execs:
            monitor.install(train_exec)

    def set_params(self, arg_params, aux_params):
        """ set parameter and aux values
        Parameters
        ----------
        arg_params : list of NDArray
            source parameter arrays
        aux_params : list of NDArray
            source aux arrays
        """

        for texec in self.execgrp.train_execs:
            texec.copy_params_from(arg_params, aux_params)

    def copy_to(self, arg_params, aux_params):
        """ Copy data from each executor to `arg_params` and `aux_params`
        Parameters
        ----------
        arg_params : list of NDArray
            target parameter arrays
        aux_params : list of NDArray
            target aux arrays
        Notes
        -----
        - This function will inplace update the NDArrays in arg_params and aux_params.
        """
        for name, block in zip(self.param_names, self.param_arrays):
            weight = sum(w.copyto(cpu()) for w in block) / len(block)
            weight.astype(arg_params[name].dtype).copyto(arg_params[name])
        for name, block in zip(self.aux_names, self.aux_arrays):
            weight = sum(w.copyto(cpu()) for w in block) / len(block)
            weight.astype(aux_params[name].dtype).copyto(aux_params[name])

    @property
    def param_arrays(self):
        """shared parameter arrays"""
        # param arrays should be shared by all executor groups
        return self.execgrp.param_arrays
    @property
    def grad_arrays(self):
        """shared gradient arrays"""
        # grad arrays should be shared by all executor groups
        return self.execgrp.grad_arrays

    @property
    def aux_arrays(self):
        """shared aux states"""
        # aux arrays are also shared by all executor groups
        return self.execgrp.aux_arrays

    def load_data_batch(self, data_batch):
        """ load data and labels into arrays """
        if self.sym_gen is not None:
            key = data_batch.bucket_key
            if key not in self.execgrp_bucket:
                # create new bucket entry
                symbol = self.sym_gen(key)
                execgrp = DataParallelExecutorGroup(symbol, self.arg_names,
                                                    self.param_names, self.ctx,
                                                    self.slices, data_batch,
                                                    shared_group=self.execgrp)
                self.execgrp_bucket[key] = execgrp

            self.curr_execgrp = self.execgrp_bucket[key]
        else:
            self.curr_execgrp = self.execgrp

        self.curr_execgrp.load_data_batch(data_batch)

    def forward(self, is_train=False):
        """run forward on the current executor"""
        self.curr_execgrp.forward(is_train=is_train)

    def backward(self):
        """run backward on the current executor"""
        self.curr_execgrp.backward()

    def update_metric(self, metric, labels):
        """update metric with the current executor"""
        self.curr_execgrp.update_metric(metric, labels)

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/ExecutorManager.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnet

import ml.dmlc.mxnet.DType.DType
import org.slf4j.{LoggerFactory, Logger}

import scala.collection.immutable.ListMap
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

/**
 * Helper class to manage multiple executors for data parallelism.
 * @author Yizhi Liu
 * @param symbol output symbol
 * @param ctx devices to run on
 * @param paramNames Name of all trainable parameters of the network.
 * @param argNames Name of all arguments of the network.
 * @param auxNames Name of all auxiliary states of the network.
 * @param trainData Training data iterator.
 * @param workLoadList The list of work load for different devices, in the same order as ctx
 * @param symGen symbol generator for bucketing
 */
private[mxnet] class DataParallelExecutorManager(private val symbol: Symbol,
                                  private val ctx: Array[Context],
                                  private[mxnet] val paramNames: IndexedSeq[String],
                                  private[mxnet] val argNames: IndexedSeq[String],
                                  private[mxnet] val auxNames: IndexedSeq[String],
                                  trainData: DataIter,
                                  private var workLoadList: Seq[Float] = null,
                                  private val symGen: SymbolGenerator = null) {
  // preparation
  private val numDevice = ctx.length
  DataParallelExecutorManager.logger.info(s"Start training with [${ctx.mkString(",")}]")

  // make sure the architecture is valid
  ExecutorManager.checkArguments(symbol)

  if (workLoadList == null) {
    workLoadList = Seq.fill(numDevice)(1f)
  }
  require(workLoadList.size == numDevice, "Invalid settings for work load.")

  private val slices = ExecutorManager.splitInputSlice(trainData.batchSize, workLoadList)

  private val paramNameSet = paramNames.toSet

  private val execGrp = new DataParallelExecutorGroup(
    symbol, argNames, paramNameSet, ctx, slices, trainData)
  private var currExecGrp: DataParallelExecutorGroup = null // this is set when data is loaded

  private val execGrpBucket: mutable.Map[AnyRef, DataParallelExecutorGroup]
    = mutable.HashMap.empty[AnyRef, DataParallelExecutorGroup]
  if (symGen != null) {
    execGrpBucket.put(trainData.defaultBucketKey, execGrp)
  }

  // shared parameter arrays
  def paramArrays: IndexedSeq[Array[NDArray]] = {
    // param arrays should be shared by all executor groups
    execGrp.paramArrays
  }

  // shared gradient arrays
  def gradArrays: IndexedSeq[Array[NDArray]] = {
    // grad arrays should be shared by all executor groups
    execGrp.gradArrays
  }

  // shared aux states
  def auxArrays: IndexedSeq[Array[NDArray]] = {
    // aux arrays are also shared by all executor groups
    execGrp.auxArrays
  }

  /**
   * Release all the executor groups.
   * The object shall never be used after it is disposed.
   */
  def dispose(): Unit = {
    execGrp.dispose()
    execGrpBucket.values.foreach(_.dispose())
  }

  // Install monitor on all executors
  def installMonitor(monitor: Monitor): Unit = {
    require(symGen == null, "Monitoring is not implemented for bucketing")
    execGrp.trainExecs.foreach(monitor.install)
  }

  /**
   * Set parameter and aux values
   * @param argParams source parameter arrays
   * @param auxParams source aux arrays
   */
  def setParams(argParams: Map[String, NDArray], auxParams: Map[String, NDArray]): Unit = {
    execGrp.trainExecs.foreach(_.copyParamsFrom(argParams, auxParams))
  }

  /**
   * Copy data from each executor to `arg_params` and `aux_params`
   * @param argParams target parameter arrays
   * @param auxParams target aux arrays
   * @note This function will inplace update the NDArrays in arg_params and aux_params.
   */
  def copyTo(argParams: Map[String, NDArray], auxParams: Map[String, NDArray]): Unit = {
    for ((name, block) <- paramNames zip paramArrays) {
      val weight = block.map(_.copyTo(Context.cpu())).reduce(_ + _) / block.length
      val typedWeight = weight.asType(argParams(name).dtype)
      typedWeight.copyTo(argParams(name))
      typedWeight.dispose()
    }
    for ((name, block) <- auxNames zip auxArrays) {
      val weight = block.map(_.copyTo(Context.cpu())).reduce(_ + _) / block.length
      val typedWeight = weight.asType(auxParams(name).dtype)
      typedWeight.copyTo(auxParams(name))
      typedWeight.dispose()
    }
  }

  // load data and labels into arrays
  def loadDataBatch(dataBatch: DataBatch): Unit = {
    currExecGrp =
      if (symGen != null) {
        val key = dataBatch.bucketKey
        require(key != null, "bucketKey must not be null for bucketing io")
        if (!execGrpBucket.contains(key)) {
          // create new bucket entry
          val sym = symGen.generate(key)
          val grp = new DataParallelExecutorGroup(sym, argNames, paramNameSet,
            ctx, slices, dataBatch, sharedGroup = execGrp)
          execGrpBucket.put(key, grp)
        }
        execGrpBucket(key)
      } else {
        execGrp
      }
    currExecGrp.loadDataBatch(dataBatch)
  }

  // run forward on the current executor
  def forward(isTrain: Boolean = false): Unit = {
    currExecGrp.forward(isTrain = isTrain)
  }

  // run backward on the current executor
  def backward(): Unit = {
    currExecGrp.backward()
  }

  // update metric with the current executor
  def updateMetric(metric: EvalMetric, labels: IndexedSeq[NDArray]): Unit = {
    currExecGrp.updateMetric(metric, labels)
  }
}

private object DataParallelExecutorManager {
  val logger: Logger = LoggerFactory.getLogger(classOf[DataParallelExecutorManager])
}

private[mxnet] object ExecutorManager {
  /**
   * Get input slice from the input shape.
   * @param batchSize The number of samples in a mini-batch.
   * @param workLoadList The list of work load for different devices, in the same order as ctx
   * @return The split slices to get a specific slice.
   * @throws IllegalArgumentException
   *         If there are two many splits such that some slice can be empty.
   */
  private[mxnet] def splitInputSlice(batchSize: Int,
                                     workLoadList: Seq[Float]): Array[(Int, Int)] = {
    val totalWorkLoad = workLoadList.sum
    val batchNumList = workLoadList.map(workLoad =>
      math.round(workLoad * batchSize / totalWorkLoad)).toArray
    val batchNumSum = batchNumList.sum
    if (batchNumSum < batchSize) {
      batchNumList(batchNumList.length-1) += batchSize - batchNumSum
    }

    val slices = ArrayBuffer.empty[(Int, Int)]
    var end = 0
    batchNumList.foreach(batchNum => {
      val begin = math.min(end, batchSize)
      end = math.min(begin + batchNum, batchSize)
      require(begin < end, "Too many slices such that some splits are empty")
      slices.append((begin, end))
    })
    slices.toArray
  }

  /**
   * Check the argument names of symbol.
   * This function checks the duplication of arguments in Symbol.
   * The check is done for feedforward net for now.
   * @param symbol The network configuration
   */
  private[mxnet] def checkArguments(symbol: Symbol): Unit = {
    val argNames = symbol.listArguments()
    require(argNames.toSet.size == argNames.length,
      "Find duplicated argument name," +
        "please make the weight name non-duplicated(using name arguments)," +
        s"arguments are $argNames")

    val auxNames = symbol.listAuxiliaryStates()
    require(auxNames.toSet.size == auxNames.length,
      "Find duplicated auxiliary param name," +
        "please make the weight name non-duplicated(using name arguments)," +
        s"arguments are $auxNames")
  }

  // Load a list of arrays into a list of arrays
  private[mxnet] def loadGeneral(data: Seq[NDArray], targets: Seq[NDArray]): Unit = {
    (data zip targets).foreach { case (dSrc, dTarget) =>
      require(dSrc.shape == dTarget.shape,
        s"src shape ${dSrc.shape} mismatch dst shape ${dTarget.shape}")
      dSrc.copyTo(dTarget)
    }
  }

  // Load a list of arrays into a list of arrays specified by slices
  private[mxnet] def loadGeneralMulti(data: Seq[NDArray],
                                      targets: Seq[Array[(Int, Int, NDArray)]]): Unit = {
    for ((src, dTargets) <- data zip targets) {
      for ((start, end, dst) <- dTargets) {
        val sliced = src.slice(start, end)
        require(sliced.shape == dst.shape,
          s"src shape ${sliced.shape} mismatch dst shape ${dst.shape}")
        sliced.copyTo(dst)
      }
    }
  }

  // Load data into sliced arrays
  private[mxnet] def loadDataMulti(batch: DataBatch,
                                   targets: Seq[Array[(Int, Int, NDArray)]]): Unit = {
    loadGeneralMulti(batch.data, targets)
  }

  private[mxnet] def loadData(batch: DataBatch, targets: Seq[NDArray]): Unit = {
    loadGeneral(batch.data, targets)
  }

  // Load label into sliced arrays
  private[mxnet] def loadLabelMulti(batch: DataBatch,
                                    targets: Seq[Array[(Int, Int, NDArray)]]): Unit = {
    loadGeneralMulti(batch.label, targets)
  }

  private[mxnet] def loadLabel(batch: DataBatch, targets: Seq[NDArray]): Unit = {
    loadGeneral(batch.label, targets)
  }

  // bind executor for bucketing, potentially sharing data with an existing executor.
  private[mxnet] def bindExec(sym: Symbol, ctx: Context, inputShapes: Map[String, Shape],
      paramNames: Set[String], needGrad: Boolean = false,
      grads: Set[String] = null, baseExec: Executor = null,
      sharedDataArrays: mutable.Map[String, NDArray] = null,
      inputTypes: ListMap[String, DType] = null) = {
    val (argShape, _, auxShape) = sym.inferShape(inputShapes)
    require(argShape != null)
    val inputTypesUpdate =
      if (inputTypes == null) {
        inputShapes.map { case (key, _) => (key, Base.MX_REAL_TYPE) }
      } else {
        inputTypes
      }
    val (argTypes, _, auxTypes) = sym.inferType(inputTypesUpdate)
    require(argTypes != null)

    val argArrays = ArrayBuffer.empty[NDArray]
    val gradArrays: mutable.Map[String, NDArray] =
      if (needGrad) mutable.HashMap.empty[String, NDArray] else null

    val argNames = sym.listArguments()

    val gradSet: Set[String] =
      if (!needGrad) {
        Set.empty[String]
      } else if (grads == null) {
        argNames.toSet -- inputShapes.keySet
      } else {
        grads
      }

    val gradReq = argNames.map { name =>
      if (gradSet.contains(name)) name -> "write"
      else name -> "null"
    }(collection.breakOut): Map[String, String]

    // create or borrow arguments and gradients
    argNames.zipWithIndex.foreach { case (name, i) =>
      if (!paramNames.contains(name)) {
        // data or label
        val argArr =
          if (sharedDataArrays != null && sharedDataArrays.contains(name)) {
            val arr = sharedDataArrays(name)
            if (arr.shape.product >= argShape(i).product) {
              // good, we can share this memory
              require(argTypes(i) == arr.dtype)
              arr.reshape(argShape(i))
            } else {
              DataParallelExecutorManager.logger.warn(
                s"bucketing: data $name has a shape ${argShape(i)}," +
                  s"which is larger than already allocated shape ${arr.shape}." +
                  "Need to re-allocate.Consider putting default_bucket_key" +
                  "to be the bucket taking the largest input for better memory sharing.")
              val zeros = NDArray.zeros(argShape(i), ctx, dtype = argTypes(i))
              // replace existing shared array because the new one is bigger
              sharedDataArrays.put(name, zeros)
              // TODO: shall we dispose the replaced array here?
              // arr.dispose()
              zeros
            }
          } else {
            val zeros = NDArray.zeros(argShape(i), ctx, dtype = argTypes(i))
            if (sharedDataArrays != null) {
              sharedDataArrays.put(name, zeros)
            }
            zeros
          }
        argArrays.append(argArr)
      } else {
        // model parameter
        val argArr =
          if (baseExec == null) {
            if (gradSet.contains(name)) {
              val gradArr = NDArray.zeros(argShape(i), ctx, dtype = argTypes(i))
              gradArrays.put(name, gradArr)
            }
            NDArray.zeros(argShape(i), ctx, dtype = argTypes(i))
          } else {
            val arr = baseExec.argDict(name)
            require(arr.shape == argShape(i))
            require(arr.dtype == argTypes(i))
            if (gradSet.contains(name)) {
              gradArrays.put(name, baseExec.gradDict(name))
            }
            arr
          }
        argArrays.append(argArr)
      }
    }
    // create or borrow aux variables
    val auxArrays =
      if (baseExec == null) {
        (auxShape zip auxTypes) map { case (s, t) =>
          NDArray.zeros(s, ctx, dtype = t)
        }
      } else {
        baseExec.auxArrays.zipWithIndex.map { case (a, i) =>
          require(auxShape(i) == a.shape)
          require(auxTypes(i) == a.dtype)
          a
        }.toSeq
      }
    sym.bind(ctx = ctx, args = argArrays.toSeq, argsGrad = gradArrays.toMap, gradsReq = gradReq,
      auxStates = auxArrays, group2ctx = null, sharedExec = baseExec)
  }
}

/**
 * A group of executors living on different devices, for data parallel.
 * @param sym The network configuration.
 * @param argNames Equals `sym.list_arguments()`
 * @param paramNames Names of all trainable parameters.
 * @param ctx List of devices for training (data parallel)
 * @param slices Describes how the data parallel splits data into different devices.
 * @param providedData training data shapes
 * @param providedLabel training label shapes
 * @param sharedGroup: DataParallelExecutorGroup
 *                   An existing executor group, if to share parameters with it.
 *
 */
private class DataParallelExecutorGroup private(sym: Symbol,
                                argNames: IndexedSeq[String], paramNames: Set[String],
                                ctx: Array[Context], private val slices: Array[(Int, Int)],
                                providedData: Map[String, Shape],
                                providedLabel: Map[String, Shape],
                                sharedGroup: DataParallelExecutorGroup)  {
  // make sure the architecture is valid
  ExecutorManager.checkArguments(sym)

  private[mxnet] val sharedDataArrays: Array[mutable.Map[String, NDArray]] =
    if (sharedGroup == null) {
      ctx.map(_ => mutable.HashMap.empty[String, NDArray])
    } else {
      sharedGroup.sharedDataArrays
    }

  private[mxnet] val dataNames = providedData.map { case (k, _) => k }.toList
  private[mxnet] val labelNames = providedLabel.map { case (k, _) => k }.toList
  private[mxnet] val auxNames = sym.listAuxiliaryStates()
  private[mxnet] val paramIdx = argNames.zipWithIndex
    .filter { case (name, i) => paramNames.contains(name) }
    .map { case (name, i) => i }
  private[mxnet] val paramNamesComb = paramIdx.map(i => argNames(i)).toSet

  private[mxnet] val trainExecs: Array[Executor] =
    ctx.zipWithIndex.map { case (ctxi, i) =>
      val dataShapes =
        (providedData ++ providedLabel) map { case (name, shape) =>
          name -> (Shape(slices(i)._2 - slices(i)._1) ++ shape.slice(1, shape.length))
        }
      val sharedExec: Executor = if (sharedGroup == null) null else sharedGroup.trainExecs(i)
      ExecutorManager.bindExec(sym, ctxi, dataShapes, paramNamesComb,
        needGrad = true, baseExec = sharedExec,
        sharedDataArrays = sharedDataArrays(i))
    }

  // data structure
  private[mxnet] val dataArrays =
    dataNames.map(name =>
      trainExecs.zipWithIndex.map { case (e, i) =>
        (slices(i)._1, slices(i)._2, e.argDict(name))
      }
    ).toIndexedSeq
  private[mxnet] val labelArrays =
    labelNames.map(name =>
      trainExecs.zipWithIndex.map { case (e, i) =>
        (slices(i)._1, slices(i)._2, e.argDict(name))
      }
    ).toIndexedSeq
  private[mxnet] val paramArrays = paramIdx.map(i =>
    trainExecs.map(e => e.argArrays(i))
  ).toIndexedSeq
  private[mxnet] val gradArrays = paramIdx.map(i =>
    trainExecs.map(e => e.gradArrays(i))
  ).toIndexedSeq
  private[mxnet] val auxArrays = (0 until auxNames.length).map(i =>
    trainExecs.map(e => e.auxArrays(i))
  )

  /**
   * A group of executors living on different devices, for data parallel
   * @param sym The network configuration.
   * @param argNames Equals `sym.list_arguments()`
   * @param paramNames Names of all trainable parameters.
   * @param ctx List of devices for training (data parallel)
   * @param slices Describes how the data parallel splits data into different devices.
   * @param trainData The dataset for training.
   *                  Loading of actual data is not necessarily needed at this stage.
   * @param sharedGroup: DataParallelExecutorGroup
   *                   An existing executor group, if to share parameters with it.
   *
   */
  def this(sym: Symbol,
      argNames: IndexedSeq[String], paramNames: Set[String],
      ctx: Array[Context], slices: Array[(Int, Int)],
      trainData: DataIter,
      sharedGroup: DataParallelExecutorGroup) {
    this(sym, argNames, paramNames, ctx, slices,
      trainData.provideData, trainData.provideLabel, sharedGroup)
  }

  def this(sym: Symbol,
           argNames: IndexedSeq[String], paramNames: Set[String],
           ctx: Array[Context], slices: Array[(Int, Int)],
           trainData: DataIter) {
    this(sym, argNames, paramNames, ctx, slices,
      trainData.provideData, trainData.provideLabel, null)
  }

  /**
   * A group of executors living on different devices, for data parallel
   * @param sym The network configuration.
   * @param argNames Equals `sym.list_arguments()`
   * @param paramNames Names of all trainable parameters.
   * @param ctx List of devices for training (data parallel)
   * @param slices Describes how the data parallel splits data into different devices.
   * @param trainData The dataset for training.
   *                  Loading of actual data is not necessarily needed at this stage.
   * @param sharedGroup: DataParallelExecutorGroup
   *                   An existing executor group, if to share parameters with it.
   *
   */
  def this(sym: Symbol,
      argNames: IndexedSeq[String], paramNames: Set[String],
      ctx: Array[Context], slices: Array[(Int, Int)],
      trainData: DataBatch,
      sharedGroup: DataParallelExecutorGroup) {
    this(sym, argNames, paramNames, ctx, slices,
      trainData.provideData, trainData.provideLabel, sharedGroup)
  }

  def this(sym: Symbol,
           argNames: IndexedSeq[String], paramNames: Set[String],
           ctx: Array[Context], slices: Array[(Int, Int)],
           trainData: DataBatch) {
    this(sym, argNames, paramNames, ctx, slices,
      trainData.provideData, trainData.provideLabel, null)
  }

  // load data and labels into arrays
  def loadDataBatch(dataBatch: DataBatch): Unit = {
    ExecutorManager.loadDataMulti(dataBatch, dataArrays)
    ExecutorManager.loadLabelMulti(dataBatch, labelArrays)
  }

  // Perform a forward pass on each executor
  def forward(isTrain: Boolean = false): Unit = {
    trainExecs.foreach(_.forward(isTrain = isTrain))
  }

  // Perform a backward pass on each executor
  def backward(): Unit = {
    trainExecs.foreach(_.backward())
  }

  // Update evaluation metric with label and current outputs
  def updateMetric(metric: EvalMetric, labels: IndexedSeq[NDArray]): Unit = {
    (trainExecs zip slices).foreach { case (texec, islice) =>
      val labelsSlice = labels.map(_.slice(islice))
      metric.update(labelsSlice, texec.outputs)
    }
  }

  /**
   * Release the related executors.
   * The object shall never be used after it is disposed.
   */
  def dispose(): Unit = {
    trainExecs.foreach(_.dispose())
  }
}


